### PR TITLE
Add libp2pctl

### PR DESCRIPTION
**Pull request by harmony-ek**
## Issue

Make it possible to control P2P connections from outside the harmony binary without having to restart or reconfigure it.  Also expose connection stats.

Since this is a disruptive operation, it requires client authentication (TLS cert auth).  Usability is to be determined/improved.

### Unit Test Coverage

TODO

### Test/Run Logs

TODO

## Operational Checklist

1. Does this PR introduce backward-incompatible changes to the on-disk data structure and/or the over-the-wire protocol? – **No**
s know about this planned change?**

8. Does this PR introduce backward-incompatible changes *NOT* related to on-disk data structure and/or over-the-wire protocol? – **No**

11. Does this PR introduce significant changes to the operational requirements of the node software, such as >20% increase in CPU, memory, and/or disk usage? – **No**

## TODO

* [ ] Add UT
* [ ] Add sample request/responses

### CODE REVIEW

File: cmd/harmony/main.go [hunk 1]
```diff
@@ -120,6 +126,12 @@ var (
 	doRevertBefore = flag.Int("do_revert_before", 0, "If the current block is less than do_revert_before, revert all blocks until (including) revert_to block")
 	revertTo       = flag.Int("revert_to", 0, "The revert will rollback all blocks until and including block number revert_to")
 	revertBeacon   = flag.Bool("revert_beacon", false, "Whether to revert beacon chain or the chain this node is assigned to")
+	// libp2p control interface
+	libp2pctlFlag         = flag.Bool("libp2pctl", false, "Enable libp2p control interface")
+	libp2pctlPortFlag     = flag.String("libp2pctl_port", "-4000", "libp2pctl port number; if prefixed with +/-, treat as offset from the -port value (default: -5000)")
```
Comment (fxfactorial) [inline cmd/harmony/main.go]: this is lot of logic in a cli flag

File: libp2pctl/server.go [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+	delete(inst.conns, ConnIDFromConn(conn).Key())
+}
+
+// Handler returns an HTTP request handler of the libp2pctl instance.
+func (inst *Instance) Handler() http.Handler {
+	return inst.rtr
+}
+
+func httpWriteJSON(w http.ResponseWriter, data interface{}) error {
+	bytes, err := json.Marshal(data)
```
Comment (fxfactorial) [inline libp2pctl/server.go]: can make shorter with: json.NewEncoder(w).Encode(data)

Comment (harmony-ek) [reply libp2pctl/server.go]: Wanted to limit truncated output case (the one-liner cannot emit HTTP 500 upon encode error).

File: cmd/harmony/main.go (hunk 1 above)
Comment (harmony-ek) [reply cmd/harmony/main.go]: Ideas?

File: libp2pctl/server.go [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+func (inst *Instance) lookupConn(r *http.Request) (network.Conn, error) {
+	vars := mux.Vars(r)
+	if vars == nil {
+		return nil, errors.New("cannot get route vars")
+	}
+	local, err := EndpointIDFromURLVar(vars["local"])
+	if err != nil {
+		return nil, errors.Wrapf(err, "invalid local endpoint %#v", vars["local"])
+	}
+	remote, err := EndpointIDFromURLVar(vars["local"])
```
Comment (gupadhyaya) [inline libp2pctl/server.go]: this should be `vars["remote"]`

**Pull request merged**